File: kvant/src/api/questionnaire.ts
import apiClient from './client';

export interface SectionProgress {
  total: number;
  completed: number;
}

export interface QuestionnaireSection {
  id: number;
  form_id: number;
  section_key: string;
  title?: string | null;
  order_index: number;
  data: Record<string, unknown>;
  is_completed: boolean;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface QuestionnaireFormSummary {
  id: number;
  guid: string;
  title: string;
  description?: string | null;
  status: string;
  current_step: number;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  progress: SectionProgress;
}

export interface QuestionnaireFormDetail
  extends Omit<QuestionnaireFormSummary, 'progress'> {
  sections: QuestionnaireSection[];
}

export interface SectionPayloadInput {
  section_key: string;
  title?: string;
  order_index?: number;
  data?: Record<string, unknown>;
  is_completed?: boolean;
  completed_at?: string;
}

export interface CreateFormPayload {
  title: string;
  description?: string;
  status?: string;
  current_step?: number;
  metadata?: Record<string, unknown>;
  sections?: SectionPayloadInput[];
}

export interface UpdateFormPayload {
  title?: string;
  description?: string;
  status?: string;
  current_step?: number;
  metadata?: Record<string, unknown>;
}

export interface UpsertSectionPayload {
  title?: string;
  order_index?: number;
  data?: Record<string, unknown>;
  is_completed?: boolean;
  completed_at?: string;
}

export const questionnaireApi = {
  async listForms(): Promise<QuestionnaireFormSummary[]> {
    const { data } = await apiClient.get<QuestionnaireFormSummary[]>('/questionnaire/forms');
    return data;
  },

  async createForm(payload: CreateFormPayload): Promise<QuestionnaireFormDetail> {
    const { data } = await apiClient.post<QuestionnaireFormDetail>('/questionnaire/forms', payload);
    return data;
  },

  async getForm(formId: number): Promise<QuestionnaireFormDetail> {
    const { data } = await apiClient.get<QuestionnaireFormDetail>(`/questionnaire/forms/${formId}`);
    return data;
  },

  async updateForm(formId: number, payload: UpdateFormPayload): Promise<QuestionnaireFormDetail> {
    const { data } = await apiClient.patch<QuestionnaireFormDetail>(`/questionnaire/forms/${formId}`, payload);
    return data;
  },

  async deleteForm(formId: number): Promise<QuestionnaireFormSummary> {
    const { data } = await apiClient.delete<QuestionnaireFormSummary>(`/questionnaire/forms/${formId}`);
    return data;
  },

  async upsertSection(
    formId: number,
    sectionKey: string,
    payload: UpsertSectionPayload,
  ): Promise<QuestionnaireSection> {
    const { data } = await apiClient.put<QuestionnaireSection>(
      `/questionnaire/forms/${formId}/sections/${sectionKey}`,
      payload,
    );
    return data;
  },
};

File: omi-dev/src/models/responses/questionnaire.py
# -*- coding: utf-8 -*-
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

__all__ = (
    "SectionProgressResponse",
    "QuestionnaireSectionResponse",
    "QuestionnaireFormSummaryResponse",
    "QuestionnaireFormDetailResponse",
)


class SectionProgressResponse(BaseModel):
    total: int = 0
    completed: int = 0


class QuestionnaireSectionResponse(BaseModel):
    id: int
    form_id: int
    section_key: str
    title: Optional[str] = None
    order_index: int
    data: Dict[str, Any] = Field(default_factory=dict)
    is_completed: bool
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class _QuestionnaireFormBase(BaseModel):
    id: int
    guid: str
    title: str
    description: Optional[str] = None
    status: str
    current_step: int
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class QuestionnaireFormSummaryResponse(_QuestionnaireFormBase):
    progress: SectionProgressResponse = Field(default_factory=SectionProgressResponse)


class QuestionnaireFormDetailResponse(_QuestionnaireFormBase):
    sections: List[QuestionnaireSectionResponse] = Field(default_factory=list)
